### PR TITLE
shuffling objective on tpu

### PR DESCRIPTION
**Pull request by ibeltagy**
- tpu doesn't support tensor indexing using another tensor, so replace this with a loss(ignore_index). This simplified the model but the data generation changed a bit
- progress bar

### CODE REVIEW

File: examples/lm_finetuning/pregenerate_training_data.py [hunk 1]
```diff
@@ -291,14 +291,12 @@ def create_instances_from_document(
                 # orig_segment_ids : [1, 1, 1, 3, 3, 0, 0, 0, 2, 2, 2, 2 ]
                 orig_segment_ids = []
                 segment_ids = []
-                for j, tokens_a in enumerate(chunks):
-                    orig_segment_ids += [chunk_ids_shuffled[j] for _ in range(len(tokens_a) + 1)]
-                    segment_ids += [j for _ in range(len(tokens_a) + 1)]
+                for j, tokens_a in enumerate(chunks_shuffled):
```
Comment (ibeltagy) [inline examples/lm_finetuning/pregenerate_training_data.py]: we didn't decide if this is a bug or not yet

File: examples/lm_finetuning/pregenerate_training_data.py [hunk 2]
```diff
@@ -291,14 +291,12 @@ def create_instances_from_document(
...
                     if j == 0:  # [CLS] token also gets an ids of 0 (similar to BERT)
                         segment_ids += [0]
-                    if chunk_ids_shuffled[j] == 0:
-                        orig_segment_ids += [0]
-
+                        orig_segment_ids += [-1]
+                    orig_segment_ids += [-1 for _ in range(len(tokens_a))] + [chunk_ids_shuffled[j]]
```
Comment (ibeltagy) [inline examples/lm_finetuning/pregenerate_training_data.py]: all `-1` follow by one label

File: pytorch_transformers/modeling_bert.py [hunk 3]
```diff
@@ -802,51 +800,24 @@ def tie_weights(self):
...
-                original_segment_ids=None):
-        # model(input_ids, segment_ids, original_segment_ids, input_mask, lm_label_ids, is_next)
+                segment_labels=None):
+
         outputs = self.bert(input_ids, position_ids=position_ids, token_type_ids=token_type_ids,
                             attention_mask=attention_mask, head_mask=head_mask)
-
         sequence_output, pooled_output = outputs[:2]
 
-        # select output locations that are [SEP] (e.g., 102) token in SciBERT vocabulary
```
Comment (ibeltagy) [inline pytorch_transformers/modeling_bert.py]: this code is replaced with predicting label for each token, then compute loss only when label != -1. This is how the masked LM objective is implemented here. This might not be the most efficient implementation, but it works well on TPU. The tf implementation actually uses the function `gather`, but the pytorch-xla implementation of `gather`  has a bug, and we waiting for it to be fixed.

File: pytorch_transformers/modeling_bert.py [hunk 4]
```diff
@@ -802,51 +800,24 @@ def tie_weights(self):
...
-        if segment_prediction_scores.shape[0] == 0:
-            # empty tensor
-            logger.error(f'There is no `[SEP]` token in the batch. Make sure preprocessing or `[SEP]` token id of {self.sep_token_id} is correct.')
-
-        # prediction_scores, seq_relationship_score = self.cls(sequence_output, pooled_output)
         # ignoring the second argument because we don't need the next-sentence-prediction
+        # we can't replace BertPreTrainingHeads with BertLMPredictionHead because we won't be able to load a pretrained model
         prediction_scores, _ = self.cls(sequence_output, pooled_output)
 
+        segment_prediction_scores = self.segment_classifier(sequence_output)  # TODO: think about using pooled_output as well. It is important for the training of BertPooler
```
Comment (ibeltagy) [inline pytorch_transformers/modeling_bert.py]: should we use `pooled_output` instead of `sequence_output` as it is the case with the `is_next` ?
